### ISSUE TITLE
[iOS] Expo-av setIsLoopingAsync Pause Between Loops

### ISSUE THREAD

**Issue by em-rich** [expo/expo#18446]
### Summary

Context:
Im implementing an application for concentration issues and relaxation. One feature of the apps is to hear sounds while concentrating in a loop.
I'm using the following configuration for the sounds:
```
export const createSound = async (inputSound) => {
    const {sound} = await Audio.Sound.createAsync({uri: inputSound.sound})
    await sound.setVolumeAsync(inputSound.volume ?? 0.5)
    await sound.setIsLoopingAsync(true)
    return sound
}
```
Actual:
The sounds start playing, when the sound finishes, one can hear a small break in-between the sounds on iOS. On Android this break cannot be heard. 

Expected:
Between two iterations of a sound no break can be heard by the user.

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

managed

### What platform(s) does this occur on?

iOS

### SDK Version (managed workflow only)

45

### Environment

System:
      OS: macOS 12.1
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 16.13.0 - ~/.nvm/versions/node/v16.13.0/bin/node
      Yarn: 1.22.17 - /opt/homebrew/bin/yarn
      npm: 8.13.2 - ~/.nvm/versions/node/v16.13.0/bin/npm
    Managers:
      CocoaPods: 1.11.2 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.0.1, iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0
    IDEs:
      Android Studio: 2021.2 AI-212.5712.43.2112.8609683
      Xcode: 13.1/13A1030d - /usr/bin/xcodebuild
    npmPackages:
      expo: ^45.0.0 => 45.0.6 
      react: 17.0.2 => 17.0.2 
      react-dom: 17.0.2 => 17.0.2 
      react-native: 0.68.2 => 0.68.2 
      react-native-web: 0.17.7 => 0.17.7 
    npmGlobalPackages:
      expo-cli: 5.3.0
    Expo Workflow: managed

### Reproducible demo

The break can be heard at 0:59-1:01.

https://snack.expo.dev/dIg5J5qd0

**Comment by IvaTsu** [comment]
I tried out the basic example from expo docs https://docs.expo.dev/versions/latest/sdk/audio/#playing-sounds with `setIsLoopingAsync(true)` and there's a slight delay before playing the file again on an iOS device (which is annoying).

**Comment by pm2877** [comment]
I am facing this issue too. Any workarounds for this?

**Comment by antondevv** [comment]
This is a problem for me as well. When I use 'setIsLoopingAsync(true)' to loop an audio file, there is a small delay/gap every time it starts over. I followed the example from expo docs https://docs.expo.dev/versions/latest/sdk/audio/#playing-sounds

**Comment by grzegorz-krzeminski** [comment]
Hey, any updates about this? it is still happening on expo 47

**Comment by timoisalive** [comment]
Yep, I'm interested in any resolutions to this one too!

**Comment by em-rich** [comment]
I finally resolved this issue by rewriting my application in Flutter and using the packages just_audio and audio_service.

**Comment by stiliajohny** [comment]
Following, I have the same issue

**Comment by mojavad** [comment]
So I've done a bit of digging on this. 

In [`EXAVPlayerData.m`](https://github.com/expo/expo/blob/d7ad395fe77f9071c27b6baf7e6df4ce706663fc/packages/expo-av/ios/EXAV/EXAVPlayerData.m#L212), the looping logic is done by setting the `actionAtItemEnd` property to `AVPlayerActionAtItemEndAdvance`, while queuing multiple instances of the asset. This effectively means it's moving through a list of the same assets to achieve a "loop".

As of iOS 10+, Swift has an [`AVPlayerLooper`](https://developer.apple.com/documentation/avfoundation/avplayerlooper) class that can achieve "seamless" loops without pauses (and would probably avoid the need to manually manage a queue here. 

It could be a fix to refactor the current looping logic to use this class instead. 

@brentvatne saw that you were on this issue before. Do you have any thoughts on this? 🙂 

Happy to try and help with this issue as it's quite important for a project I'm currently working on.

**Comment by stiliajohny** [comment]
@mojavad 
What a legend, this is great 
I wish I could help! 

i think `expo-av` is pretty much all we have when it comes to sound player, react-native-sound is not as good

**Comment by lukewlms** [comment]
I'm not aware of a fix yet for this issue which I'd summarize as "unpredictable gap when looping sound"

**Comment by lukewlms** [comment]
For now we've worked around this issue by creating a very long track that has the sound looping (for a metronome), but this significantly increases app size.

**Comment by em-rich** [comment]
yes, we have done that in the beginning too. You should think about storing
your sounds on a server and load it from there.

**Comment by Peege151** [comment]
Just discovered this behavior as well...  for now I'm going to export longer versions of the audio as well... not ideal but the issue has been open for over a year and I'm not able to help with this directly.  It would be nice to have a seamless loop ability though

**Comment by nikocraft** [comment]
I solved this issue by choosing only to release on Android, but it would be nice to have a proper technical solution, so I'm bumping this thread! Expo team, could you take a look at this?

**Comment by OlivierCo** [comment]
Please can you try to make a PR @mojavad ?

**Comment by OlivierCo** [comment]
@nikocraft i don't get why you thumb down my comment. As @mojavad could fix it.

**Comment by nikocraft** [comment]
@OlivierCo you downvoted everyone else so it was response to that, to show a sign of appreciation on your downvotes! :D A nobody gonna fix this issue, if they were, they already would have! Have a nice day

**Comment by mojavad** [comment]
@OlivierCo - got some bad news, I'm afraid...

I spent a fair bit of time trying to fix this, but it's turned out to be a lot more complicated than expected. 

The code for the looping is heavily intertwined with a bunch of other callbacks that are used for other parts of the playback functionality. Despite spending several hours trying to debug and make sense of it, I couldn't get it to work where it wouldn't break something else 😔

I ultimately gave up on it... 

Given the effort to build out a separate `expo-video` package that's modern and written fully in Swift, I think it might be better to just create an Expo Module for audio playback and drop to the native player and implement looping with `AVPlayerLooper`...

**Comment by allredj** [comment]
Hey!
I'd love it if this could be fixed.
We are working around this issue by running two concurrent soundtracks and fading them in and out of each other, but it's obviously not ideal.

Is there a timeline for the `AVPlayerLooper` solution?
How can I help?

**Comment by nikocraft** [comment]
> ###
> ** [allredj](/allredj)**

There is no timeline, we are stuck with what we have right now.

**Comment by MdAyman7** [comment]
Any workaround on this?

**Comment by smachi** [comment]
We are experiencing the same issue.
The react-native-sound library works fine with loops in iOS but not in Android.
And expo-av is working fine in Android but not in iOS...

**Comment by RobertoGallego** [comment]
any solution on this ?

**Comment by jacksweeney24** [comment]
im still having this same issue

**Comment by devsphere-apps** [comment]
+1

**Comment by OskarJuszczyk** [comment]
+1, still happening on expo-audio

**Comment by 227Faddi** [comment]
+1
